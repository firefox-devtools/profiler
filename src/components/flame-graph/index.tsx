/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { StackSettings } from '../shared/StackSettings';
import { TransformNavigator } from '../shared/TransformNavigator';
import { MaybeFlameGraph } from './MaybeFlameGraph';

const FlameGraphView = () => (
  <div
    className="flameGraph"
    id="flame-graph-tab"
    role="tabpanel"
    aria-labelledby="flame-graph-tab-button"
  >
    <StackSettings hideInvertCallstack={true} />
    <TransformNavigator />
    <MaybeFlameGraph />
  </div>
);

export const FlameGraph = FlameGraphView;
