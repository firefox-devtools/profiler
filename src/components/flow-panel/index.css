/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.flowPanel {
  display: flex;
  flex: 1;
  flex-flow: column nowrap;
}

.markerChartCanvas {
  border-top: 1px solid var(--grey-30);
}
