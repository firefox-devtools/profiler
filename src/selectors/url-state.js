/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';
import { ensureExists, getFirstItemFromSet } from '../utils/flow';
import { urlFromState } from '../app-logic/url-handling';
import { getThreadsKey } from '../profile-logic/profile-data';
import { stringsToMarkerRegExps } from '../profile-logic/marker-data';
import { getProfileNameFromZipPath } from 'firefox-profiler/profile-logic/zip-files';
import { SYMBOL_SERVER_URL } from '../app-logic/constants';
import { splitSearchString, stringsToRegExp } from '../utils/string';
import { isLocalURL } from '../utils/url';

import type {
  ThreadIndex,
  Pid,
  TransformStack,
  Action,
  TimelineType,
  DataSource,
  ImplementationFilter,
  CallTreeSummaryStrategy,
  UrlState,
  TimelineTrackOrganization,
  Selector,
  DangerousSelectorWithArguments,
  StartEndRange,
  TrackIndex,
  ThreadsKey,
  ProfileSpecificUrlState,
  FullProfileSpecificUrlState,
  ActiveTabSpecificProfileUrlState,
  NativeSymbolInfo,
  TabID,
} from 'firefox-profiler/types';

import type { TabSlug } from '../app-logic/tabs-handling';
import type { MarkerRegExps } from '../profile-logic/marker-data';

import urlStateReducer from '../reducers/url-state';
import { formatMetaInfoString } from '../profile-logic/profile-metainfo';

/**
 * Various simple selectors into the UrlState.
 */
export const getUrlState: Selector<UrlState> = (state): UrlState =>
  state.urlState;
export const getProfileSpecificState: Selector<ProfileSpecificUrlState> = (
  state
) => getUrlState(state).profileSpecific;
export const getFullProfileSpecificState: Selector<
  FullProfileSpecificUrlState,
> = (state) => getProfileSpecificState(state).full;
export const getActiveTabProfileSpecificState: Selector<
  ActiveTabSpecificProfileUrlState,
> = (state) => getProfileSpecificState(state).activeTab;

export const getDataSource: Selector<DataSource> = (state) =>
  getUrlState(state).dataSource;
export const getHash: Selector<string> = (state) => getUrlState(state).hash;
export const getProfileUrl: Selector<string> = (state) =>
  getUrlState(state).profileUrl;
export const getProfilesToCompare: Selector<string[] | null> = (state) =>
  getUrlState(state).profilesToCompare;
export const getProfileNameFromUrl: Selector<string | null> = (state) =>
  getUrlState(state).profileName;
export const getAllCommittedRanges: Selector<StartEndRange[]> = (state) =>
  getProfileSpecificState(state).committedRanges;
export const getImplementationFilter: Selector<ImplementationFilter> = (
  state
) => getProfileSpecificState(state).implementation;
export const getLastSelectedCallTreeSummaryStrategy: Selector<
  CallTreeSummaryStrategy,
> = (state) =>
  getProfileSpecificState(state).lastSelectedCallTreeSummaryStrategy;
export const getShowUserTimings: Selector<boolean> = (state) =>
  getProfileSpecificState(state).showUserTimings;
export const getSourceViewFile: Selector<string | null> = (state) =>
  getProfileSpecificState(state).sourceView.sourceFile;
export const getSourceViewScrollGeneration: Selector<number> = (state) =>
  getProfileSpecificState(state).sourceView.scrollGeneration;
export const getAssemblyViewIsOpen: Selector<boolean> = (state) =>
  getProfileSpecificState(state).assemblyView.isOpen;
export const getAssemblyViewNativeSymbol: Selector<NativeSymbolInfo | null> = (
  state
) => getProfileSpecificState(state).assemblyView.nativeSymbol;
export const getAssemblyViewScrollGeneration: Selector<number> = (state) =>
  getProfileSpecificState(state).assemblyView.scrollGeneration;
export const getShowJsTracerSummary: Selector<boolean> = (state) =>
  getFullProfileSpecificState(state).showJsTracerSummary;
export const getTimelineTrackOrganization: Selector<
  TimelineTrackOrganization,
> = (state) => getUrlState(state).timelineTrackOrganization;

/**
 * Active tab specific url state selectors
 */
export const getIsActiveTabResourcesPanelOpen: Selector<boolean> = (state) =>
  getActiveTabProfileSpecificState(state).isResourcesPanelOpen;

/**
 * Raw search strings, before any splitting has been performed.
 */
export const getCurrentSearchString: Selector<string> = (state) =>
  getProfileSpecificState(state).callTreeSearchString;
export const getMarkersSearchString: Selector<string> = (state) =>
  getProfileSpecificState(state).markersSearchString;
export const getNetworkSearchString: Selector<string> = (state) =>
  getProfileSpecificState(state).networkSearchString;
export const getSelectedTab: Selector<TabSlug> = (state) =>
  getUrlState(state).selectedTab;
export const getInvertCallstack: Selector<boolean> = (state) =>
  getSelectedTab(state) === 'calltree' &&
  getProfileSpecificState(state).invertCallstack;

export const getSelectedThreadIndexesOrNull: Selector<
  Set<ThreadIndex> | null,
> = (state) => getProfileSpecificState(state).selectedThreads;
export const getSelectedThreadIndexes: Selector<Set<ThreadIndex>> = (state) =>
  ensureExists(
    getSelectedThreadIndexesOrNull(state),
    'Attempted to get a thread index before a profile was loaded.'
  );
export const getSelectedThreadsKey: Selector<ThreadsKey> = (state) =>
  getThreadsKey(getSelectedThreadIndexes(state));

/**
 * This selector is temporary for a migration to multiple selected thread indexes.
 */
export const getFirstSelectedThreadIndex: Selector<ThreadIndex> = (state) =>
  ensureExists(
    getFirstItemFromSet(getSelectedThreadIndexes(state)),
    'Expected to find at least one thread index in the selected thread indexes'
  );
export const getTimelineType: Selector<TimelineType> = (state) =>
  getProfileSpecificState(state).timelineType;

/**
 * Simple selectors for tracks and track order.
 */
export const getLegacyThreadOrder: Selector<ThreadIndex[] | null> = (state) =>
  getFullProfileSpecificState(state).legacyThreadOrder;
export const getLegacyHiddenThreads: Selector<ThreadIndex[] | null> = (state) =>
  getFullProfileSpecificState(state).legacyHiddenThreads;
export const getGlobalTrackOrder: Selector<TrackIndex[]> = (state) =>
  getFullProfileSpecificState(state).globalTrackOrder;
export const getHiddenGlobalTracks: Selector<Set<TrackIndex>> = (state) =>
  getFullProfileSpecificState(state).hiddenGlobalTracks;
export const getHiddenLocalTracksByPid: Selector<Map<Pid, Set<TrackIndex>>> = (
  state
) => getFullProfileSpecificState(state).hiddenLocalTracksByPid;
export const getLocalTrackOrderByPid: Selector<Map<Pid, TrackIndex[]>> = (
  state
) => getFullProfileSpecificState(state).localTrackOrderByPid;
export const getTabFilter: Selector<TabID | null> = (state) =>
  getFullProfileSpecificState(state).tabFilter;
export const hasTabFilter: Selector<boolean> = (state) =>
  getTabFilter(state) !== null;

/**
 * This selector does a simple lookup in the set of hidden tracks for a PID, and ensures
 * that a TrackIndex is selected correctly. This makes it easier to avoid doing null
 * checks everywhere.
 */
export const getHiddenLocalTracks: DangerousSelectorWithArguments<
  Set<TrackIndex>,
  Pid,
> = (state, pid) =>
  ensureExists(
    getHiddenLocalTracksByPid(state).get(pid),
    'Unable to get the hidden tracks from the given pid'
  );

/**
 * This selector gets the local track order for a PID, and ensures that one is
 * selected correctly. This makes it easier to avoid doing null checks everywhere.
 */
export const getLocalTrackOrder: DangerousSelectorWithArguments<
  TrackIndex[],
  Pid,
> = (state, pid) =>
  ensureExists(
    getLocalTrackOrderByPid(state).get(pid),
    'Unable to get the track order from the given pid'
  );

/**
 * Search strings filter a thread to only samples that match the strings.
 */
export const getSearchStrings: Selector<string[] | null> = createSelector(
  getCurrentSearchString,
  splitSearchString
);

export const getMarkersSearchStrings: Selector<string[] | null> =
  createSelector(getMarkersSearchString, splitSearchString);

export const getNetworkSearchStrings: Selector<string[] | null> =
  createSelector(getNetworkSearchString, splitSearchString);

/**
 * A RegExp can be used for searching and filtering the thread's samples.
 */
export const getSearchStringsAsRegExp: Selector<RegExp | null> = createSelector(
  getSearchStrings,
  stringsToRegExp
);

export const getMarkersSearchStringsAsRegExp: Selector<MarkerRegExps | null> =
  createSelector(getMarkersSearchStrings, stringsToMarkerRegExps);

export const getNetworkSearchStringsAsRegExp: Selector<MarkerRegExps | null> =
  createSelector(getNetworkSearchStrings, stringsToMarkerRegExps);

// Pre-allocate an array to help with strict equality tests in the selectors.
const EMPTY_TRANSFORM_STACK = [];

export const getTransformStack: DangerousSelectorWithArguments<
  TransformStack,
  ThreadsKey,
> = (state, threadsKey) => {
  return (
    getProfileSpecificState(state).transforms[threadsKey] ||
    EMPTY_TRANSFORM_STACK
  );
};

export const getIsBottomBoxOpen: Selector<boolean> = (state) => {
  const tab = getSelectedTab(state);
  return getProfileSpecificState(state).isBottomBoxOpenPerPanel[tab];
};

/**
 * The URL predictor is used to generate a link for an uploaded profile, to predict
 * what the URL will be.
 */
export const getUrlPredictor: Selector<(Action | Action[]) => string> =
  createSelector(
    getUrlState,
    (oldUrlState: UrlState) => (actionOrActionList: Action | Action[]) => {
      const actionList: Action[] = Array.isArray(actionOrActionList)
        ? actionOrActionList
        : [actionOrActionList];
      const newUrlState = actionList.reduce(urlStateReducer, oldUrlState);
      return urlFromState(newUrlState);
    }
  );

/**
 * Get the current path for a zip file that is being used.
 */
export const getPathInZipFileFromUrl: Selector<string | null> = (state) =>
  getUrlState(state).pathInZipFile;

/**
 * Get a short formatted string that represents the meta info of the current profile.
 */
export const getFormattedMetaInfoString: Selector<string | null> = (state) => {
  // Avoid circular dependencies by selected the profile meta manually.
  const { profile } = state.profileView;
  if (!profile) {
    return null;
  }

  return formatMetaInfoString(profile.meta);
};

/**
 * Get just the file name from the zip file path, if it exists.
 */
export const getFileNameInZipFilePath: Selector<string | null> = createSelector(
  getPathInZipFileFromUrl,
  (pathInZipFile) => {
    if (pathInZipFile) {
      return getProfileNameFromZipPath(pathInZipFile);
    }
    return null;
  }
);

/**
 * Get a profile name that can be used as a short identifier for the profile. This
 * will be displayed in the UI, and will be used as a default for a profile name
 * if none is currently set. If none is set, then a series of strategies will be
 * used to select a default one.
 */
export const getProfileNameWithDefault: Selector<string> = createSelector(
  getProfileNameFromUrl,
  getFileNameInZipFilePath,
  getFormattedMetaInfoString,
  (profileNameFromUrl, fileNameInZipFilePath, formattedMetaInfoString) => {
    // Always prefer a manually set name.
    if (profileNameFromUrl) {
      return profileNameFromUrl;
    }
    // Next, try and use a path from the zip file.
    if (fileNameInZipFilePath) {
      return fileNameInZipFilePath;
    }

    // Finally return a generic string describing the type of profile.
    if (!formattedMetaInfoString) {
      return 'Untitled profile';
    }
    return formattedMetaInfoString;
  }
);

/**
 * Determines the profile name used to store in the IndexedDB by default.
 */
export const getProfileNameForStorage: Selector<string> = createSelector(
  getProfileNameFromUrl,
  getFileNameInZipFilePath,
  (profileNameFromUrl, fileNameInZipFilePath) => {
    // Always prefer a manually set name.
    if (profileNameFromUrl) {
      return profileNameFromUrl;
    }
    // Next, try and use a path from the zip file.
    if (fileNameInZipFilePath) {
      return fileNameInZipFilePath;
    }

    // Finally, return a blank string.
    return '';
  }
);

function _shouldAllowSymbolServerUrl(symbolServerUrl) {
  try {
    const url = new URL(symbolServerUrl);
    if (isLocalURL(url)) {
      return true;
    }

    // For non-localhost hosts, we require HTTPS.
    const otherAllowedHostnames = [
      'symbols.mozilla.org',
      'symbolication.services.mozilla.com',
    ];
    if (!otherAllowedHostnames.includes(url.hostname)) {
      console.error(
        `The symbol server URL was not in the list of allowed domains. Rejecting ${symbolServerUrl} and defaulting to ${SYMBOL_SERVER_URL}.`
      );
      return false;
    }
    if (url.protocol !== 'https:') {
      console.error(
        `HTTPS is required for non-localhost symbol servers. Rejecting ${symbolServerUrl} and defaulting to ${SYMBOL_SERVER_URL}.`
      );
      return false;
    }
    return true;
  } catch (error) {
    console.error(
      `The symbol server URL was not valid. Rejecting ${symbolServerUrl} and defaulting to ${SYMBOL_SERVER_URL}.`,
      error
    );
    return false;
  }
}

export const getSymbolServerUrl: Selector<string> = createSelector(
  getUrlState,
  ({ symbolServerUrl }) => {
    if (!symbolServerUrl || !_shouldAllowSymbolServerUrl(symbolServerUrl)) {
      return SYMBOL_SERVER_URL;
    }

    // Strip off the trailing slash, if present.
    if (symbolServerUrl.endsWith('/')) {
      return symbolServerUrl.slice(0, -1);
    }
    return symbolServerUrl;
  }
);
