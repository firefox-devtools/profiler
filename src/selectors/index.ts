/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
export * from './app';
export * from './l10n';
export * from './per-thread';
export * from './profile';
export * from './url-state';
export * from './icons';
export * from './publish';
export * from './zipped-profiles';
export * from './cpu';
export * from './code';

import * as app from './app';
import {
  selectedThreadSelectors as selectedThread,
  selectedNodeSelectors as selectedNode,
} from './per-thread';
import * as profile from './profile';
import * as urlState from './url-state';
import * as icons from './icons';
import * as publish from './publish';
import * as zippedProfiles from './zipped-profiles';
import * as l10n from './l10n';
import * as cpu from './cpu';
import * as code from './code';

const _selectorsForConsole = {
  app,
  profile,
  urlState,
  icons,
  publish,
  zippedProfiles,
  selectedThread,
  selectedNode,
  l10n,
  cpu,
  code,
};

// Exports require explicit typing. Deduce the type with typeof.
export const selectorsForConsole: typeof _selectorsForConsole =
  _selectorsForConsole;
