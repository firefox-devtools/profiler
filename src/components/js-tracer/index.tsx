/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
import * as React from 'react';
import explicitConnect from 'firefox-profiler/utils/connect';
import { JsTracerChart } from './Chart';
import { JsTracerSettings } from './Settings';
import { JsTracerEmptyReasons } from './EmptyReasons';

import { getProfile } from 'firefox-profiler/selectors/profile';
import { selectedThreadSelectors } from 'firefox-profiler/selectors/per-thread';
import {
  getShowJsTracerSummary,
  getSelectedThreadsKey,
} from 'firefox-profiler/selectors/url-state';
import { updatePreviewSelection } from 'firefox-profiler/actions/profile-view';

import type {
  Profile,
  JsTracerTable,
  ThreadsKey,
} from 'firefox-profiler/types';
import type { ConnectedProps } from 'firefox-profiler/utils/connect';

import './index.css';

type DispatchProps = {
  readonly updatePreviewSelection: typeof updatePreviewSelection;
};

type StateProps = {
  readonly profile: Profile;
  readonly threadsKey: ThreadsKey;
  readonly jsTracerTable: JsTracerTable | null;
  readonly showJsTracerSummary: boolean;
};

type Props = ConnectedProps<{}, StateProps, DispatchProps>;

class JsTracerImpl extends React.PureComponent<Props> {
  override render() {
    const { profile, jsTracerTable, showJsTracerSummary, threadsKey } =
      this.props;
    return (
      <div className="jsTracer">
        {jsTracerTable === null || jsTracerTable.events.length === 0 ? (
          <JsTracerEmptyReasons />
        ) : (
          <>
            <JsTracerSettings />
            <JsTracerChart
              profile={profile}
              jsTracerTable={jsTracerTable}
              showJsTracerSummary={showJsTracerSummary}
              threadsKey={threadsKey}
            />
          </>
        )}
      </div>
    );
  }
}

export const JsTracer = explicitConnect<{}, StateProps, DispatchProps>({
  mapStateToProps: (state) => {
    return {
      profile: getProfile(state),
      threadsKey: getSelectedThreadsKey(state),
      jsTracerTable: selectedThreadSelectors.getJsTracerTable(state),
      showJsTracerSummary: getShowJsTracerSummary(state),
    };
  },
  mapDispatchToProps: { updatePreviewSelection },
  component: JsTracerImpl,
});
