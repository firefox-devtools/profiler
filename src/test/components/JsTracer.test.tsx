/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { Provider } from 'react-redux';

import { render, screen } from 'firefox-profiler/test/fixtures/testing-library';
import {
  TIMELINE_MARGIN_LEFT,
  TIMELINE_MARGIN_RIGHT,
} from '../../app-logic/constants';
import { JsTracer } from '../../components/js-tracer';
import { getShowJsTracerSummary } from '../../selectors/url-state';

import {
  autoMockCanvasContext,
  flushDrawLog,
} from '../fixtures/mocks/canvas-context';
import { mockRaf } from '../fixtures/mocks/request-animation-frame';
import { storeWithProfile } from '../fixtures/stores';
import {
  addRootOverlayElement,
  removeRootOverlayElement,
  fireFullClick,
} from '../fixtures/utils';
import {
  getProfileWithJsTracerEvents,
  type TestDefinedJsTracerEvent,
} from '../fixtures/profiles/processed-profile';
import { autoMockElementSize } from '../fixtures/mocks/element-size';

jest.useFakeTimers();

const GRAPH_BASE_WIDTH = 200;
const GRAPH_WIDTH =
  GRAPH_BASE_WIDTH + TIMELINE_MARGIN_LEFT + TIMELINE_MARGIN_RIGHT;
const GRAPH_HEIGHT = 300;

describe('StackChart', function () {
  autoMockCanvasContext();
  autoMockElementSize({ width: GRAPH_WIDTH, height: GRAPH_HEIGHT });
  beforeEach(addRootOverlayElement);
  afterEach(removeRootOverlayElement);

  function setup({
    skipLoadingScreen,
    events,
  }: {
    skipLoadingScreen: boolean;
    events: TestDefinedJsTracerEvent[];
  }) {
    const flushRafCalls = mockRaf();

    const profile = getProfileWithJsTracerEvents(events);

    const store = storeWithProfile(profile);
    const { getState, dispatch } = store;

    const renderResult = render(
      <Provider store={store}>
        <JsTracer />
      </Provider>
    );
    const { container } = renderResult;

    if (skipLoadingScreen) {
      flushRafCalls();
    }

    const getJsTracerChartCanvas = () => container.querySelector('canvas');
    const getChangeJsTracerSummaryCheckbox = () =>
      screen.getByText(/Show only self time/);

    return {
      ...renderResult,
      dispatch,
      getState,
      flushRafCalls,
      getJsTracerChartCanvas,
      getChangeJsTracerSummaryCheckbox,
    };
  }

  const simpleTracerEvents: TestDefinedJsTracerEvent[] = [
    ['https://mozilla.org', 0, 20],
    ['Interpreter', 1, 19],
    ['IonMonkey', 2, 18],
  ];

  it('only computes the chart after the first tick', () => {
    const { getJsTracerChartCanvas, flushRafCalls } = setup({
      skipLoadingScreen: false,
      events: simpleTracerEvents,
    });
    expect(getJsTracerChartCanvas()).toBeFalsy();
    flushRafCalls();
    expect(getJsTracerChartCanvas()).toBeTruthy();
  });

  it('matches the snapshot for the loading screen', () => {
    const { container } = setup({
      skipLoadingScreen: false,
      events: simpleTracerEvents,
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches the snapshot for empty reasons screen', () => {
    const { container } = setup({
      skipLoadingScreen: true,
      events: [],
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches the snapshot for the chart', () => {
    const { container } = setup({
      skipLoadingScreen: true,
      events: simpleTracerEvents,
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches the snapshot a simple chart render', () => {
    setup({
      skipLoadingScreen: true,
      events: simpleTracerEvents,
    });
    expect(flushDrawLog()).toMatchSnapshot();
  });

  it('can change to a summary view', function () {
    const { getChangeJsTracerSummaryCheckbox, getState } = setup({
      skipLoadingScreen: true,
      events: simpleTracerEvents,
    });
    expect(getShowJsTracerSummary(getState())).toEqual(false);
    fireFullClick(getChangeJsTracerSummaryCheckbox());
    expect(getShowJsTracerSummary(getState())).toEqual(true);
  });

  it('matches the snapshot for an inverted draw call', function () {
    const { getChangeJsTracerSummaryCheckbox } = setup({
      skipLoadingScreen: true,
      events: simpleTracerEvents,
    });
    fireFullClick(getChangeJsTracerSummaryCheckbox());
    expect(flushDrawLog()).toMatchSnapshot();
  });
});
