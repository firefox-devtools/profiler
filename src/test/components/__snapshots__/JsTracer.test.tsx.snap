// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`StackChart matches the snapshot a simple chart render 1`] = `
Array [
  Array [
    "set font",
    "10px sans-serif",
  ],
  Array [
    "measureText",
    "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.()< /:-_",
  ],
  Array [
    "measureText",
    "…",
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    0,
    0,
    365,
    300,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    150,
    1,
    200,
    13,
  ],
  Array [
    "measureText",
    "https://mozilla.org",
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "https://mozilla.org",
    153,
    11,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    160,
    17,
    179,
    13,
  ],
  Array [
    "measureText",
    "Interpreter",
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "Interpreter",
    163,
    27,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    339,
    18,
    0.9999999999999745,
    11,
  ],
  Array [
    "fillRect",
    170,
    33,
    160,
    13,
  ],
  Array [
    "measureText",
    "IonMonkey",
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "IonMonkey",
    173,
    43,
  ],
  Array [
    "set fillStyle",
    "#ededf0",
  ],
  Array [
    "fillRect",
    149,
    0,
    1,
    300,
  ],
  Array [
    "fillRect",
    0,
    15,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    31,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    47,
    365,
    1,
  ],
  Array [
    "set fillStyle",
    "#000000",
  ],
  Array [
    "measureText",
    "Tracing Information",
  ],
  Array [
    "fillText",
    "Tracing Information",
    5,
    11,
  ],
  Array [
    "set font",
    "10px sans-serif",
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    0,
    0,
    365,
    300,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    150,
    1,
    200,
    13,
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "https://mozilla.org",
    153,
    11,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    160,
    17,
    179,
    13,
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "Interpreter",
    163,
    27,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    339,
    18,
    0.9999999999999745,
    11,
  ],
  Array [
    "fillRect",
    170,
    33,
    160,
    13,
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "IonMonkey",
    173,
    43,
  ],
  Array [
    "set fillStyle",
    "#ededf0",
  ],
  Array [
    "fillRect",
    149,
    0,
    1,
    300,
  ],
  Array [
    "fillRect",
    0,
    15,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    31,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    47,
    365,
    1,
  ],
  Array [
    "set fillStyle",
    "#000000",
  ],
  Array [
    "fillText",
    "Tracing Information",
    5,
    11,
  ],
]
`;

exports[`StackChart matches the snapshot for an inverted draw call 1`] = `
Array [
  Array [
    "set font",
    "10px sans-serif",
  ],
  Array [
    "measureText",
    "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.()< /:-_",
  ],
  Array [
    "measureText",
    "…",
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    0,
    0,
    365,
    300,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    150,
    1,
    200,
    13,
  ],
  Array [
    "measureText",
    "https://mozilla.org",
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "https://mozilla.org",
    153,
    11,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    160,
    17,
    179,
    13,
  ],
  Array [
    "measureText",
    "Interpreter",
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "Interpreter",
    163,
    27,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    339,
    18,
    0.9999999999999745,
    11,
  ],
  Array [
    "fillRect",
    170,
    33,
    160,
    13,
  ],
  Array [
    "measureText",
    "IonMonkey",
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "IonMonkey",
    173,
    43,
  ],
  Array [
    "set fillStyle",
    "#ededf0",
  ],
  Array [
    "fillRect",
    149,
    0,
    1,
    300,
  ],
  Array [
    "fillRect",
    0,
    15,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    31,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    47,
    365,
    1,
  ],
  Array [
    "set fillStyle",
    "#000000",
  ],
  Array [
    "measureText",
    "Tracing Information",
  ],
  Array [
    "fillText",
    "Tracing Information",
    5,
    11,
  ],
  Array [
    "set font",
    "10px sans-serif",
  ],
  Array [
    "set fillStyle",
    "#ffffff",
  ],
  Array [
    "fillRect",
    0,
    0,
    365,
    300,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    150,
    1,
    200,
    13,
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "https://mozilla.org",
    153,
    11,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    160,
    17,
    179,
    13,
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "Interpreter",
    163,
    27,
  ],
  Array [
    "set fillStyle",
    "#45a1ff",
  ],
  Array [
    "fillRect",
    339,
    18,
    0.9999999999999745,
    11,
  ],
  Array [
    "fillRect",
    170,
    33,
    160,
    13,
  ],
  Array [
    "set fillStyle",
    "white",
  ],
  Array [
    "fillText",
    "IonMonkey",
    173,
    43,
  ],
  Array [
    "set fillStyle",
    "#ededf0",
  ],
  Array [
    "fillRect",
    149,
    0,
    1,
    300,
  ],
  Array [
    "fillRect",
    0,
    15,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    31,
    365,
    1,
  ],
  Array [
    "fillRect",
    0,
    47,
    365,
    1,
  ],
  Array [
    "set fillStyle",
    "#000000",
  ],
  Array [
    "fillText",
    "Tracing Information",
    5,
    11,
  ],
]
`;

exports[`StackChart matches the snapshot for empty reasons screen 1`] = `
<div
  class="jsTracer"
>
  <div
    class="EmptyReasons"
  >
    <h3>
      The 
      js-tracer
       is empty for “
      Empty
      ”
    </h3>
    <p>
      This thread contains no JS tracer information.
    </p>
  </div>
</div>
`;

exports[`StackChart matches the snapshot for the chart 1`] = `
<div
  class="jsTracer"
>
  <div
    class="jsTracerSettings"
  >
    <ul
      class="jsTracerSettingsList"
    >
      <li
        class="jsTracerSettingsListItem"
      >
        <label
          class="jsTracerSettingsLabel"
        >
          <input
            class="jsTracerSettingsCheckbox"
            type="checkbox"
          />
          <span
            title="Show only the time spent in a call node, ignoring its children."
          >
            Show only self time
          </span>
        </label>
      </li>
    </ul>
  </div>
  <div
    class="chartViewport"
    tabindex="0"
  >
    <div>
      <canvas
        class="chartCanvas jsTracerCanvasFadeIn jsTracerCanvas jsTracerCanvasDrawn"
        height="300"
        style="width: 365px; height: 300px;"
        width="365"
      />
    </div>
    <div
      class="chartViewportScroll hidden"
    >
      Zoom Chart:
      <kbd
        class="chartViewportScrollKbd"
      >
        Ctrl + Scroll
      </kbd>
      or
      <kbd
        class="chartViewportScrollKbd"
      >
        Shift + Scroll
      </kbd>
    </div>
  </div>
</div>
`;

exports[`StackChart matches the snapshot for the loading screen 1`] = `
<div
  class="jsTracer"
>
  <div
    class="jsTracerSettings"
  >
    <ul
      class="jsTracerSettingsList"
    >
      <li
        class="jsTracerSettingsListItem"
      >
        <label
          class="jsTracerSettingsLabel"
        >
          <input
            class="jsTracerSettingsCheckbox"
            type="checkbox"
          />
          <span
            title="Show only the time spent in a call node, ignoring its children."
          >
            Show only self time
          </span>
        </label>
      </li>
    </ul>
  </div>
  <div
    class="jsTracerLoader"
  >
    Re-constructing tracing information from
     
    3
     events. This might take a moment.
  </div>
</div>
`;
